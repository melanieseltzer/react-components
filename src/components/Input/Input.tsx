import React from 'react';
import styled from 'styled-components';

export interface Input extends React.InputHTMLAttributes<HTMLInputElement> {
  background?: string;
  border?: number;
  color?: string;
  width?: string;
}

interface Props extends Input {
  float?: boolean;
}

const Input = (props: Input): JSX.Element => <StyledInput {...props} />;

export default Input;

const StyledInput = styled.input`
  background: ${(props: Props) =>
    props.background ? props.background : 'none'};
  border: 0;
  border-bottom: ${(props: Props) => `${props.border ? props.border : 1}px solid
    ${props.color ? props.color : `#ccc`}`};
  font-size: 1rem;
  height: auto;
  transition: all 0.2s;
  padding: 0.8rem 0 0.5rem 0;
  ${(props: Props) => props.width && `width: ${props.width}`};
  ${(props: Props) =>
    props.float &&
    `
      /* normal label size (when placeholder is there but hidden aka not focused) */
      :placeholder-shown + label {
        transform-origin: left bottom;
        transform: translate(0, 0.4rem) scale(1);
      }
      /* when not focused, hide the placeholder */
      :placeholder-shown:not(:focus)::placeholder,
      :placeholder-shown:not(:focus)::-webkit-input-placeholder {
        opacity: 0;
        transition: inherit;
      }
      /* show the placeholder only when input is focused */
      :focus::placeholder,
      :focus::-webkit-input-placeholder {
        color: #626262;
        opacity: 1;
      }
      /* shrink the label when focused and also when text has been entered */
      :not(:placeholder-shown) + label,
      :focus + label {
        padding-top: 0.2rem;
        transform: translate(0, -0.7rem) scale(0.8);
      }
    `}
`;
