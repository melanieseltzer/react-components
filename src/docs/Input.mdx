---
name: Input
menu: Components
route: /input
---

import { Playground, Props } from 'docz';
import Input from '../components/Input/Input';
import Label from '../components/Label/Label';
import Flex from '../components/Flex/Flex';

# Input

An HTML input component. Should be used with [Label](/label).

## Props

<Props of={Input} />

## Basic usage

Normal border:

<Playground>
  <Input
    className="custom-class"
    id="name"
    name="name"
    placeholder="John Smith"
    width="100%"
    border={1}
  />
</Playground>

Border bottom:

<Playground>
  <Input
    className="custom-class"
    id="name2"
    name="name2"
    placeholder="John Smith"
    color="tomato"
    border={1}
    borderBottom
    width="100%"
  />
</Playground>
