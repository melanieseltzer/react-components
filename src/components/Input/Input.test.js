import React from 'react';
import { render, cleanup } from 'react-testing-library';
import 'jest-styled-components';
import Input from '.';

afterEach(cleanup);

describe('Input component', () => {
  it('matches the snapshot', () => {
    const { container } = render(<Input />);
    expect(container).toMatchSnapshot();
  });

  it('should render with a class of test', () => {
    const { container } = render(<Input className="test" />);
    expect(container.firstChild.classList.contains('test')).toBe(true);
  });

  it('should render with float styles if float prop is passed', () => {
    const { container } = render(<Input float />);
    expect(container.firstChild).toHaveStyleRule('opacity', '0', {
      modifier: `
        :placeholder-shown:not(:focus)::placeholder
      `
    });
  });
});
