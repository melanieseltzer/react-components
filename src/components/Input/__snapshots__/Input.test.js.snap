// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input component matches the snapshot 1`] = `
.c0 {
  background: none;
  border: 0;
  font-size: 16px;
  -webkit-transition: all 0.2s;
  transition: all 0.2s;
  padding: 0.5rem;
}

<div>
  <input
    color="#ccc"
    class="c0"
  />
</div>
`;
